official/coin_build/man/man3/SoWWWAnchor.3
.TH "SoWWWAnchor" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoWWWAnchor \- The \fBSoWWWAnchor\fP class adds URL callbacks to the highlighted geometry\&.
.PP
In addition to highlighting geometry under the cursor, the application programmer can set callbacks\&. It is possible to set one callback for picking, the fetch callback, and one callback for highlighting\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/nodes/SoWWWAnchor\&.h>\fP
.PP
Inherits \fBSoLocateHighlight\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBMapping\fP { \fBNONE\fP, \fBPOINT\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "\fBSoWWWAnchor\fP (void)"
.br
.ti -1c
.RI "void \fBsetFullURLName\fP (const \fBSbString\fP &url)"
.br
.ti -1c
.RI "const \fBSbString\fP & \fBgetFullURLName\fP (void)"
.br
.ti -1c
.RI "virtual void \fBhandleEvent\fP (\fBSoHandleEventAction\fP *action)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBsetFetchURLCallBack\fP (SoWWWAnchorCB *f, void *userData)"
.br
.ti -1c
.RI "static void \fBsetHighlightURLCallBack\fP (SoWWWAnchorCB *f, void *userData)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBSoSFString\fP \fBname\fP"
.br
.ti -1c
.RI "\fBSoSFString\fP \fBdescription\fP"
.br
.ti -1c
.RI "\fBSoSFEnum\fP \fBmap\fP"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual const \fBSoFieldData\fP * \fBgetFieldData\fP (void) const"
.br
.ti -1c
.RI "virtual \fB~SoWWWAnchor\fP ()"
.br
.ti -1c
.RI "virtual void \fBredrawHighlighted\fP (\fBSoAction\fP *act, SbBool isNowHighlighting)"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetFieldDataPtr\fP (void)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
The \fBSoWWWAnchor\fP class adds URL callbacks to the highlighted geometry\&.
.PP
In addition to highlighting geometry under the cursor, the application programmer can set callbacks\&. It is possible to set one callback for picking, the fetch callback, and one callback for highlighting\&. 


.PP
.nf
#Inventor V2.1 ascii

WWWAnchor {
   name "http://www.coin3d.org/Coin/egg.iv"
   description "Easter Egg"

   Separator {
      Transform { scaleFactor 0.8 1.2 0.8 }
      Sphere { }
   }
}

.fi
.PP
.PP
\fBFILE FORMAT/DEFAULTS:\fP 
.PP
.nf
WWWAnchor {
    renderCaching AUTO
    boundingBoxCaching AUTO
    renderCulling AUTO
    pickCulling AUTO
    color 0\&.3 0\&.3 0\&.3
    style EMISSIVE
    mode AUTO
    name "<Undefined URL>"
    description ""
    map NONE
}

.fi
.PP
.PP
\fBSince:\fP
.RS 4
Inventor 2\&.1 
.RE
.PP

.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBSoWWWAnchor::Mapping\fP"
Enum that says how a picked node's position should be mapped to the URL\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fINONE \fP\fP
The position of the picked node is not mapped to the URL\&. 
.TP
\fB\fIPOINT \fP\fP
The position of the picked node is mapped to the URL as object space coordinates, adding a parameter string to the end of the URL\&. To assure that the URL works with all browsers, the coordinates are divided by commas sent as the hex representation\&.
.PP
If a model by the name of sim\&.wrl resided at www\&.coin3d\&.org and the picked point had the coordinates [1\&.5, 10, 6\&.77], the resulting URL would be 'http://www\&.coin3d\&.org/sim\&.wrl?1\&.5%2c10%2c6\&.77'\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoWWWAnchor::SoWWWAnchor (void)"
Constructor\&. 
.SS "SoWWWAnchor::~SoWWWAnchor ()\fC [protected]\fP, \fC [virtual]\fP"
Destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP SoWWWAnchor::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. Usage example:
.PP
.PP
.nf
void foo(SoNode * node)
{
  if (node->getTypeId() == SoFile::getClassTypeId()) {
    SoFile * filenode = (SoFile *)node;  // safe downward cast, knows the type
  }
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new nodes, engines, nodekits, draggers or others: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through for instance \fBInventor/nodes/SoSubNode\&.h\fP (SO_NODE_INIT_CLASS and SO_NODE_CONSTRUCTOR for node classes), \fBInventor/engines/SoSubEngine\&.h\fP (for engine classes) and so on\&.
.PP
For more information on writing Coin extensions, see the class documentation of the toplevel superclasses for the various class groups\&. 
.PP
Reimplemented from \fBSoLocateHighlight\fP\&.
.SS "const \fBSoFieldData\fP * SoWWWAnchor::getFieldData (void) const\fC [protected]\fP, \fC [virtual]\fP"
Returns a pointer to the class-wide field data storage object for this instance\&. If no fields are present, returns \fCNULL\fP\&. 
.PP
Reimplemented from \fBSoLocateHighlight\fP\&.
.SS "void SoWWWAnchor::setFullURLName (const \fBSbString\fP & url)"
Sets the full URL to \fIurl\fP\&. If this is set, this URL will be used in callbacks instead of the URL set in \fBSoWWWAnchor::name\fP\&.
.PP
\fBSee also:\fP
.RS 4
\fBSoWWWAnchor::getFullURLName()\fP 
.RE
.PP

.SS "const \fBSbString\fP & SoWWWAnchor::getFullURLName (void)"
Returns the full URL if it's set by \fBSoWWWAnchor::setFullURLName()\fP\&. Otherwise the contents of \fBSoWWWAnchor::name\fP is returned\&.
.PP
\fBSee also:\fP
.RS 4
\fBSoWWWAnchor::setFullURLName()\fP 
.RE
.PP

.SS "void SoWWWAnchor::handleEvent (\fBSoHandleEventAction\fP * action)\fC [virtual]\fP"
Action method for \fBSoHandleEventAction\fP\&.
.PP
Inspects the event data from \fIaction\fP, and processes it if it is something which this node should react to\&.
.PP
Nodes influencing relevant state variables for how event handling is done also overrides this method\&. 
.PP
Reimplemented from \fBSoLocateHighlight\fP\&.
.SS "void SoWWWAnchor::setFetchURLCallBack (SoWWWAnchorCB * f, void * userData)\fC [static]\fP"
Sets the callback function \fIf\fP that is called when a \fBSoWWWAnchor\fP node is clicked on\&. This callback can among other things be used to provide a browser with the URL of this node\&.
.PP
The callback will be called with the URL, \fIuserData\fP and a pointer to this node as arguments\&. 
.SS "void SoWWWAnchor::setHighlightURLCallBack (SoWWWAnchorCB * f, void * userData)\fC [static]\fP"
Sets the callback function \fIf\fP that is called when a \fBSoWWWAnchor\fP node is highlighted\&. This callback can among other things be used to provide the user with a visual clue on which URL the node points to, for example by showing the URL as a string\&.
.PP
The callback will be called with the URL, \fIuserData\fP and a pointer to this node as arguments\&. 
.SS "void SoWWWAnchor::redrawHighlighted (\fBSoAction\fP * act, SbBool isNowHighlighting)\fC [protected]\fP, \fC [virtual]\fP"
Calls the highlight callback set up with \fBSoWWWAnchor::setHighlightURLCallBack()\fP\&. 
.PP
Reimplemented from \fBSoLocateHighlight\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBSoSFString\fP SoWWWAnchor::name"
The name of the URL which the anchor points to\&. 
.SS "\fBSoSFString\fP SoWWWAnchor::description"
The description of the URL\&. 
.SS "\fBSoSFEnum\fP SoWWWAnchor::map"
Enum describing how a node's position should be mapped to the URL\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/nodes/SoWWWAnchor.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoWWWAnchor SoWWWAnchor.h Inventor/nodes/SoWWWAnchor.h
  \brief The SoWWWAnchor class adds URL callbacks to the highlighted geometry.
  \ingroup nodes

  In addition to highlighting geometry under the cursor, the application
  programmer can set callbacks. It is possible to set one callback for
  picking, the fetch callback, and one callback for highlighting.

  \verbatim
  #Inventor V2.1 ascii
  
  WWWAnchor {
     name "http://www.coin3d.org/Coin/egg.iv"
     description "Easter Egg"
  
     Separator {
        Transform { scaleFactor 0.8 1.2 0.8 }
        Sphere { }
     }
  }
  \endverbatim

  <b>FILE FORMAT/DEFAULTS:</b>
  \code
    WWWAnchor {
        renderCaching AUTO
        boundingBoxCaching AUTO
        renderCulling AUTO
        pickCulling AUTO
        color 0.3 0.3 0.3
        style EMISSIVE
        mode AUTO
        name "<Undefined URL>"
        description ""
        map NONE
    }
  \endcode

  \since Inventor 2.1
*/


/*! \file SoWWWAnchor.h */
#include <Inventor/nodes/SoWWWAnchor.h>

#include <Inventor/events/SoMouseButtonEvent.h>
#include <Inventor/actions/SoHandleEventAction.h>
#include <Inventor/SoPickedPoint.h>

#include "nodes/SoSubNodeP.h"
#include "coindefs.h"
#include "tidbitsp.h"


/*!
  \enum SoWWWAnchor::Mapping
  Enum that says how a picked node's position should be mapped to the URL.
*/
/*!
  \var SoWWWAnchor::Mapping SoWWWAnchor::NONE
  The position of the picked node is not mapped to the URL.
*/
/*!
  \var SoWWWAnchor::Mapping SoWWWAnchor::POINT
  The position of the picked node is mapped to the URL as object space
  coordinates, adding a parameter string to the end of the URL. To
  assure that the URL works with all browsers, the coordinates are
  divided by commas sent as the hex representation.

  If a model by the name of sim.wrl resided at www.coin3d.org and the
  picked point had the coordinates [1.5, 10, 6.77], the resulting URL
  would be "http://www.coin3d.org/sim.wrl?1.5%2c10%2c6.77".
*/


/*!
  \var SoSFString SoWWWAnchor::name
  The name of the URL which the anchor points to.
*/
/*!
  \var SoSFString SoWWWAnchor::description
  The description of the URL.
*/
/*!
  \var SoSFEnum SoWWWAnchor::map
  Enum describing how a node's position should be mapped to the URL.
*/

// *************************************************************************

#ifndef DOXYGEN_SKIP_THIS

class SoWWWAnchorP {
 public:
  SoWWWAnchorP(SoWWWAnchor * ownerptr) {
    this->owner = ownerptr;
    this->fullname = "";
  }
  SoWWWAnchor * owner;
  SbString fullname;

  static SoWWWAnchorCB * fetchfunc;
  static void * fetchdata;
  static SoWWWAnchorCB * highlightfunc;
  static void * highlightdata;

  static void atexit_cleanup() {
    SoWWWAnchorP::fetchfunc = nullptr;
    SoWWWAnchorP::fetchdata = nullptr;
    SoWWWAnchorP::highlightfunc = nullptr;
    SoWWWAnchorP::highlightdata = nullptr;
  }
};

// static members
SoWWWAnchorCB * SoWWWAnchorP::fetchfunc = nullptr;
void * SoWWWAnchorP::fetchdata = nullptr;
SoWWWAnchorCB * SoWWWAnchorP::highlightfunc = nullptr;
void * SoWWWAnchorP::highlightdata = nullptr;

#endif // DOXYGEN_SKIP_THIS

SO_NODE_SOURCE(SoWWWAnchor);

#define PRIVATE(p) (p->pimpl)

/*!
  Constructor.
*/
SoWWWAnchor::SoWWWAnchor()
{
  SO_NODE_INTERNAL_CONSTRUCTOR(SoWWWAnchor);

  PRIVATE(this) = new SoWWWAnchorP(this);

  SO_NODE_ADD_FIELD(name, ("<Undefined URL>"));
  SO_NODE_ADD_FIELD(description, (""));
  SO_NODE_ADD_FIELD(map, (NONE));

  SO_NODE_DEFINE_ENUM_VALUE(Map, NONE);
  SO_NODE_DEFINE_ENUM_VALUE(Map, POINT);
  SO_NODE_SET_SF_ENUM_TYPE(map, Map);
}

/*!
  Destructor.
*/
SoWWWAnchor::~SoWWWAnchor()
{
  delete PRIVATE(this);
}

// doc in super
void
SoWWWAnchor::initClass(void)
{
  SO_NODE_INTERNAL_INIT_CLASS(SoWWWAnchor, SO_FROM_INVENTOR_2_1|SoNode::VRML1);
  coin_atexit((coin_atexit_f*)SoWWWAnchorP::atexit_cleanup, CC_ATEXIT_NORMAL);
}


/*!
  Sets the full URL to \a url. If this is set, this URL will be used in
  callbacks instead of the URL set in SoWWWAnchor::name.

  \sa SoWWWAnchor::getFullURLName()
 */
void
SoWWWAnchor::setFullURLName(const SbString & url)
{
  PRIVATE(this)->fullname = url;
}

/*!
  Returns the full URL if it's set by
  SoWWWAnchor::setFullURLName(). Otherwise the contents of
  SoWWWAnchor::name is returned.

  \sa SoWWWAnchor::setFullURLName()
 */
const SbString &
SoWWWAnchor::getFullURLName(void)
{
  if (PRIVATE(this)->fullname.getLength() > 0) {
    return PRIVATE(this)->fullname;
  }

  return this->name.getValue();
}

// documented in superclass
void
SoWWWAnchor::handleEvent(SoHandleEventAction * action)
{
  const SoEvent * event = action->getEvent();
  if (event->isOfType(SoMouseButtonEvent::getClassTypeId()) &&
      SoWWWAnchorP::fetchfunc) {
    const SoMouseButtonEvent * mbevent = (SoMouseButtonEvent*)event;
    if (SoMouseButtonEvent::isButtonPressEvent(mbevent,
                                               SoMouseButtonEvent::BUTTON1)) {
      SbString s = this->getFullURLName();
      if (this->map.getValue() == POINT) {
        const SoPickedPoint * pp = action->getPickedPoint();
        const SbVec3f point = pp->getObjectPoint(nullptr);
        SbString temp;
        temp.sprintf("?%g%%2c%g%%2c%g", point[0], point[1], point[2]);
        s.operator+=(temp);
      }

      SoWWWAnchorP::fetchfunc(s, SoWWWAnchorP::fetchdata, this);
    }
  }
  inherited::handleEvent(action);
}

/*!
  Sets the callback function \a f that is called when a SoWWWAnchor node is
  clicked on. This callback can among other things be used to provide a
  browser with the URL of this node.

  The callback will be called with the URL, \a userData and a pointer to
  this node as arguments.
 */
void
SoWWWAnchor::setFetchURLCallBack(SoWWWAnchorCB * f, void * userData)
{
  SoWWWAnchorP::fetchfunc = f;
  SoWWWAnchorP::fetchdata = userData;
}

/*!
  Sets the callback function \a f that is called when a SoWWWAnchor node
  is highlighted. This callback can among other things be used to provide
  the user with a visual clue on which URL the node points to, for example
  by showing the URL as a string.

  The callback will be called with the URL, \a userData and a pointer to
  this node as arguments.
 */
void
SoWWWAnchor::setHighlightURLCallBack(SoWWWAnchorCB * f, void * userData)
{
  SoWWWAnchorP::highlightfunc = f;
  SoWWWAnchorP::highlightdata = userData;
}

/*!
  Calls the highlight callback set up with
  SoWWWAnchor::setHighlightURLCallBack().
*/
void
SoWWWAnchor::redrawHighlighted(SoAction * act, bool isNowHighlighting)
{
  inherited::redrawHighlighted(act, isNowHighlighting);

  if (SoWWWAnchorP::highlightfunc) {
    SbString s = this->getFullURLName();
    SoWWWAnchorP::highlightfunc(s, SoWWWAnchorP::highlightdata, this);
  }
}

#undef PRIVATE
